File: scripts/test-api-performance.js
#!/usr/bin/env node
/**
 * API Performance Testing Script
 * Validates P95 latency < 200ms for /api/opportunities
 */

const { performance } = require('perf_hooks');
const fs = require('fs');

const API_BASE = process.env.API_BASE || 'http://localhost:4173';
const ITERATIONS = parseInt(process.env.PERF_ITERATIONS || '100');
const P95_THRESHOLD = parseInt(process.env.P95_THRESHOLD || '200'); // ms

async function measureAPIPerformance() {
  const results = [];
  const warmupRuns = 10;
  
  console.log(`🔥 Warming up API with ${warmupRuns} requests...`);
  
  // Warmup phase
  for (let i = 0; i < warmupRuns; i++) {
    try {
      await fetch(`${API_BASE}/api/opportunities?page=1&limit=50`);
    } catch (error) {
      console.warn(`Warmup request ${i + 1} failed:`, error.message);
    }
  }
  
  console.log(`📊 Running ${ITERATIONS} performance measurements...`);
  
  // Measurement phase
  for (let i = 0; i < ITERATIONS; i++) {
    const start = performance.now();
    
    try {
      const response = await fetch(`${API_BASE}/api/opportunities?page=${(i % 5) + 1}&limit=100`);
      const end = performance.now();
      const duration = end - start;
      
      if (response.ok) {
        results.push({
          iteration: i + 1,
          duration,
          status: response.status,
          timestamp: new Date().toISOString()
        });
      } else {
        console.warn(`Request ${i + 1} failed with status:`, response.status);
      }
    } catch (error) {
      console.error(`Request ${i + 1} failed:`, error.message);
    }
    
    // Progress indicator
    if ((i + 1) % 20 === 0) {
      console.log(`Progress: ${i + 1}/${ITERATIONS} (${Math.round((i + 1) / ITERATIONS * 100)}%)`);
    }
  }
  
  if (results.length === 0) {
    throw new Error('No successful API requests recorded');
  }
  
  // Calculate statistics
  const durations = results.map(r => r.duration).sort((a, b) => a - b);
  const stats = {
    total_requests: results.length,
    min_duration: Math.min(...durations),
    max_duration: Math.max(...durations),
    avg_duration: durations.reduce((sum, d) => sum + d, 0) / durations.length,
    p50_duration: durations[Math.floor(durations.length * 0.5)],
    p95_duration: durations[Math.floor(durations.length * 0.95)],
    p99_duration: durations[Math.floor(durations.length * 0.99)],
    threshold: P95_THRESHOLD,
    success_rate: (results.length / ITERATIONS) * 100
  };
  
  // Save detailed results
  const report = {
    test_name: 'API Performance Test',
    timestamp: new Date().toISOString(),
    configuration: {
      api_base: API_BASE,
      iterations: ITERATIONS,
      p95_threshold: P95_THRESHOLD
    },
    statistics: stats,
    raw_data: results
  };
  
  // Ensure reports directory exists
  if (!fs.existsSync('reports')) {
    fs.mkdirSync('reports', { recursive: true });
  }
  
  const filename = `reports/performance-api-${Date.now()}.json`;
  fs.writeFileSync(filename, JSON.stringify(report, null, 2));
  
  // Console output
  console.log('\n📈 API Performance Results:');
  console.log(`Average Response Time: ${stats.avg_duration.toFixed(2)}ms`);
  console.log(`P50 Response Time: ${stats.p50_duration.toFixed(2)}ms`);
  console.log(`P95 Response Time: ${stats.p95_duration.toFixed(2)}ms`);
  console.log(`P99 Response Time: ${stats.p99_duration.toFixed(2)}ms`);
  console.log(`Success Rate: ${stats.success_rate.toFixed(2)}%`);
  console.log(`Report saved: ${filename}`);
  
  // Validation
  const passed = stats.p95_duration <= P95_THRESHOLD && stats.success_rate >= 95;
  
  if (passed) {
    console.log(`✅ PASS: P95 latency ${stats.p95_duration.toFixed(2)}ms <= ${P95_THRESHOLD}ms threshold`);
    process.exit(0);
  } else {
    console.log(`❌ FAIL: P95 latency ${stats.p95_duration.toFixed(2)}ms > ${P95_THRESHOLD}ms threshold`);
    process.exit(1);
  }
}

// Run the test
measureAPIPerformance().catch(error => {
  console.error('❌ Performance test failed:', error.message);
  process.exit(1);
});

File: scripts/test_api_performance.py
#!/usr/bin/env python3
"""
API Performance Testing Script - Phase 1 CI Gate
Validates P95 latency < 200ms for /api/opportunities
"""

import asyncio
import aiohttp
import time
import json
import statistics
import os
from typing import List, Dict, Any
from pathlib import Path

API_BASE = os.getenv('API_BASE', 'http://localhost:4173')
P95_THRESHOLD = int(os.getenv('P95_THRESHOLD', '200'))  # ms
ITERATIONS = int(os.getenv('PERF_ITERATIONS', '100'))

async def measure_api_performance() -> Dict[str, Any]:
    """Measure API performance with concurrent requests"""
    print(f"🔥 Starting API performance test - P95 threshold: {P95_THRESHOLD}ms")
    
    async with aiohttp.ClientSession() as session:
        # Warmup phase
        print("Warming up API...")
        for i in range(10):
            try:
                async with session.get(f"{API_BASE}/api/opportunities?limit=50") as resp:
                    await resp.text()
            except Exception as e:
                print(f"Warmup request {i+1} failed: {e}")
        
        # Measurement phase
        print(f"📊 Running {ITERATIONS} performance measurements...")
        durations = []
        
        tasks = []
        for i in range(ITERATIONS):
            tasks.append(measure_single_request(session, i))
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Filter successful results
        for result in results:
            if isinstance(result, dict) and 'duration' in result:
                durations.append(result['duration'])
        
        if not durations:
            raise Exception("No successful API requests recorded")
        
        # Calculate statistics
        durations.sort()
        stats = {
            'total_requests': len(durations),
            'min_duration': min(durations),
            'max_duration': max(durations), 
            'avg_duration': statistics.mean(durations),
            'p50_duration': statistics.median(durations),
            'p95_duration': durations[int(len(durations) * 0.95)],
            'p99_duration': durations[int(len(durations) * 0.99)],
            'threshold': P95_THRESHOLD,
            'success_rate': (len(durations) / ITERATIONS) * 100
        }
        
        # Save report
        report = {
            'test_name': 'API Performance Test',
            'timestamp': time.time(),
            'configuration': {
                'api_base': API_BASE,
                'iterations': ITERATIONS,
                'p95_threshold': P95_THRESHOLD
            },
            'statistics': stats
        }
        
        # Ensure reports directory exists
        Path('reports').mkdir(exist_ok=True)
        
        with open(f'reports/api-performance-{int(time.time())}.json', 'w') as f:
            json.dump(report, f, indent=2)
        
        # Console output
        print(f"\n📈 API Performance Results:")
        print(f"Average Response Time: {stats['avg_duration']:.2f}ms")
        print(f"P95 Response Time: {stats['p95_duration']:.2f}ms")
        print(f"Success Rate: {stats['success_rate']:.2f}%")
        
        # Validation
        passed = stats['p95_duration'] <= P95_THRESHOLD and stats['success_rate'] >= 95
        
        if passed:
            print(f"✅ PASS: P95 latency {stats['p95_duration']:.2f}ms <= {P95_THRESHOLD}ms")
            return stats
        else:
            print(f"❌ FAIL: P95 latency {stats['p95_duration']:.2f}ms > {P95_THRESHOLD}ms")
            exit(1)

async def measure_single_request(session: aiohttp.ClientSession, iteration: int) -> Dict[str, Any]:
    """Measure a single API request"""
    start_time = time.time()
    
    try:
        endpoint = f"/api/opportunities?page={(iteration % 5) + 1}&limit=100"
        async with session.get(f"{API_BASE}{endpoint}") as response:
            await response.text()
            end_time = time.time()
            duration = (end_time - start_time) * 1000  # Convert to ms
            
            if response.status == 200:
                return {
                    'iteration': iteration + 1,
                    'duration': duration,
                    'status': response.status
                }
            else:
                print(f"Request {iteration + 1} failed with status: {response.status}")
                return {}
    except Exception as e:
        print(f"Request {iteration + 1} failed: {e}")
        return {}

if __name__ == "__main__":
    asyncio.run(measure_api_performance())